chapters/chapter_7/section_1.md

```diff
@@ -0,0 +1,271 @@
+#define AFFILE_PARSER_H_INCLUDED
+
+#include "cfg-parser.h"
+#include "driver.h"
+```
+
+Here we declare the `CfgParser`. It is defined in the source file for the parser (`affile-parser.c`).
+```
+extern CfgParser affile_parser;
+```
+
+This macro function declares the Bison lex and error functions for this module's parser.
+```
+CFG_PARSER_DECLARE_LEXER_BINDING(affile_, LogDriver **)
+
+#endif
+```
+
+### `affile-parser.c`
+```
+#include "driver.h"
+#include "cfg-parser.h"
+```
+
+`affile-grammar.h` is the header for the Bison parser implementation file. The implementation file and header are generated by Bison at compile-time, based on the module's grammar file (`affile-grammar.ym` in this case).
+```
+#include "affile-grammar.h"
+
+extern int affile_debug;
+```
+
+This is the Bison parser that comes from the generated implementation file.
+```
+int affile_parse(CfgLexer *lexer, LogDriver **instance, gpointer arg);
+```
+
+The lexer for the module's parser needs a list of keywords to recognize, so we create an array of `CfgLexerKeyword` objects. These are defined by the string to recognize, along with the corresponding token's numerical code. The tokens are defined in the `*-grammar.ym` file and Bison makes their numerical codes availible through a macro of the same name, so that is what we use. The array is null-terminated.
+
+The keywords of modules include those that declare the use of the plugins (the first block here), as well as those that set the options for the plugins (the other blocks here).
+```
+static CfgLexerKeyword affile_keywords[] =
+{
+  { "file",               KW_FILE },
+  { "fifo",               KW_PIPE },
+  { "pipe",               KW_PIPE },
+  { "stdin",              KW_STDIN },
+
+  { "wildcard_file",      KW_WILDCARD_FILE },
+  { "base_dir",           KW_BASE_DIR },
+  { "filename_pattern",   KW_FILENAME_PATTERN },
+  { "recursive",          KW_RECURSIVE },
+  { "max_files",          KW_MAX_FILES },
+  { "monitor_method",     KW_MONITOR_METHOD },
+
+  { "fsync",              KW_FSYNC },
+  { "remove_if_older",    KW_OVERWRITE_IF_OLDER, KWS_OBSOLETE, "overwrite_if_older" },
+  { "overwrite_if_older", KW_OVERWRITE_IF_OLDER },
+  { "follow_freq",        KW_FOLLOW_FREQ },
+  { "multi_line_mode",    KW_MULTI_LINE_MODE  },
+  { "multi_line_prefix",  KW_MULTI_LINE_PREFIX },
+  { "multi_line_garbage", KW_MULTI_LINE_GARBAGE },
+  { "multi_line_suffix",  KW_MULTI_LINE_GARBAGE },
+  { NULL }
+};
+```
+
+Now we can define our `CfgParser` (boilerplate).
+```
+CfgParser affile_parser =
+{
+#if SYSLOG_NG_ENABLE_DEBUG
+  .debug_flag = &affile_debug,
+#endif
+  .name = "affile",
+  .keywords = affile_keywords,
+  .parse = (gint (*)(CfgLexer *, gpointer *, gpointer)) affile_parse,
+  .cleanup = (void (*)(gpointer)) log_pipe_unref,
+};
+```
+
+Lastly we need to call this macro function to define our lex and error functions.
+```
+CFG_PARSER_IMPLEMENT_LEXER_BINDING(affile_, LogDriver **)
+```
+
+## OOP in C
```

Comment:
As I've started to work on the fixes for this part, I am begining to realise that the entire section is probably best left out, since the vast amount of small details will only serve to confuse and overwhelm the reader. When I was first trying to comprehend the codebase, I found it mostly easy to pick up how structs were used as classes, and learned the nuances over time, and so I think this topic is best understood by just reading and interacting with the code.

So I will probably change this section into a small note about the fact that structs are used to represent classes, but not go into the details of inheritance or type casting, or static methods and the like, and just say that the reader will learn this stuff easiest through exposure to the code.

Reply:
Sounds good to me.